keps/sig-api-machinery/3920-finish-feature-flags/README.md

```diff
@@ -0,0 +1,1015 @@
+  the feature enablement doesn't break the whole cluster at once when there is a
+  problem.
+
+### Non-Goals
+
+* The feature system will not permit passing configuration, since that's already
+  done in a variety of ways today. Therefore only features that don't take
+  config can be dynamic. Maybe this can be addressed in a follow up.
+
+## Proposal
+
+### High-level design
+
+First we will discuss all the pieces with minimal detail. Probably this section
+will leave you with questions which hopefully are answered in the "detailed
+design" section.
+
+Below we will talk about various actors in the system:
+* Clients: Clients exist today, they read feature gate values from the command
+  line, and check various features to see if they are enabled.
+* Devs: developers write implementations for features.
+* APIs: Kubernetes APIs.
+* "Servers": Today there isn't really a server or anything like it involved in
+  features.
+* Users: end users of a Kubernetes cluster without special permissions.
+* Cluster admins: users with admin rights over the cluster, but possibly not
+  access to the command line of binaries like apiserver, especially on managed
+  clouds.
+* Platform admins: this person has command line access to binaries. They
+  probably don't use the cluster but are responsible for it staying healthy.
+
+#### Client Flow
+
+Let's call the current use of feature gates the "client flow". Things that are
+already handled correctly by the existing client flow we will try not to change.
+
+The main thing we want to add to the client flow is reporting of the feature's
+uses: did the client check the feature state, and what value did it find? The
+necessity of this will be clear when we discuss the "server flow" later.
+
+To make our improvements to the client flow, we will add a new implementation of
+the existing k8s.io/component-base/featuregate package's MutableFeatureGate
+type. It will require post-construction initialization, so that it can write the
+uses to the api; to be a drop-in replacement, it can't require an api client
+when it is constructed, because that happens before the command line is parsed /
+config files are read.
+
+Our new client will use the existing MutableFeatureGate to read the command
+line, so perhaps it should be thought of as a layer.
+
+When we update this KEP to solve the secondary goals, we will add here a
+mechanism for reading feature on/off values from the API -- that's not easy, due
+to the startup order.
+
+#### API
+
+Obviously, we will add an API object, Feature (details below). Features will be
+namespaced; this will permit all Kubernetes features to be in the kube-system
+namespace, and third parties can use another namespace.
+
+The Feature object will record, per-feature, its configured desired value,
+default value, uses, and current version. Most importantly, it will record the
+current state of the feature. There will be at least four states: On, Off,
+TurningOn, and TurningOff.
```

Comment:
I think we would benefit a lot if you would actually provide the API structure here.

What I'm wondering about is how we represent the state (and how we generally track it) if the feature-gate is used by multiple components (let's say apiserver, scheduler and kubelets). Then technically the feature is on only when all of them are actually enabled.
Also - what we're doing with cases with 1 (or small percent) of kuebelets (or any other node components) are not changing their flag?

Reply:
So we can know the status of the FGs by `kubectl get features`. 
- https://github.com/kubernetes/kubernetes/pull/116887 is not needed then.
- https://github.com/kubernetes/website/issues/25645: a proposal to add a history of feature-gate feed in JSON/table like the CVE feed. This may help to do some cleanup or check.

Reply:
I was going to save the API type for the detailed design section, which I haven't written yet. But you can look at the prototype [here](https://github.com/lavalamp/kubernetes/blob/prototype-gates/staging/src/k8s.io/api/features/v1alpha1/types.go). There's some stuff about it in the prototype that is wrong and needs to change, and there's other stuff that is right but counterintuitive, so reader beware, it will take some mental effort to tell the difference if you follow that link :)